tooling-docker/src/main/java/org/jboss/eap/qe/ts/common/docker/Docker.java

```diff
@@ -156,54 +259,104 @@ public boolean isRunning() throws Exception {
+        } catch (InterruptedException e) {
+            throw new ContainerStopException("A thread was interrupted while waiting for its termination!", e);
+        }
     }
 
-    public void kill() throws Exception {
-        System.out.println(Ansi.ansi().reset().a("Killing container ").fgCyan().a(name).reset()
+    /**
+     * Kill this docker container using docker command. Be aware that there might occur a situation when the docker
+     * command might fail. This might be caused for example by reaching file descriptors limit in system.
+     * 
+     * @throws ContainerKillException thrown when the kill command fails.
+     */
+    public void kill() throws ContainerKillException {
+        this.out.println(Ansi.ansi().reset().a("Killing container ").fgCyan().a(this.name).reset()
                 .a(" with ID ").fgYellow().a(uuid).reset());
 
-        new ProcessBuilder()
-                .command("docker", "kill", uuid)
-                .start()
-                .waitFor(10, TimeUnit.SECONDS);
-        terminateThreadPools();
-        removeDockerContainer();
+        try {
+            runDockerCommand("kill", this.uuid);
+        } catch (DockerCommandException e) {
+            throw new ContainerKillException("Failed to kill the container '" + this.uuid + "'!", e);
+        }
+
+        try {
+            terminatePrintingThread();
+        } catch (InterruptedException e) {
+            throw new ContainerKillException("Interrupted when waiting for printer thread termination!", e);
+        }
     }
 
-    private void terminateThreadPools() throws Exception {
-        outputPrinter.shutdown();
-        outputPrinter.awaitTermination(10, TimeUnit.SECONDS);
+    private void terminatePrintingThread() throws InterruptedException {
+        if (outputPrintingThread != null) {
+            outputPrintingThread.shutdown();
+            outputPrintingThread.awaitTermination(10, TimeUnit.SECONDS);
+        }
     }
 
-    private void removeDockerContainer() throws Exception {
-        new ProcessBuilder()
-                .command("docker", "rm", uuid)
-                .start()
-                .waitFor(10, TimeUnit.SECONDS);
+    @Override
+    public void remove() throws ContainerRemoveException {
+        try {
+            runDockerCommand("rm", this.uuid);
+        } catch (DockerCommandException e) {
+            throw new ContainerRemoveException("Failed to remove the container '" + this.uuid + "'!", e);
+        }
+    }
+
+    private Process runDockerCommand(final String... commandArguments) throws DockerCommandException {
+        final String dockerCommand = "docker";
+        final List<String> cmd = new ArrayList<>();
+        cmd.add(dockerCommand);
+        Collections.addAll(cmd, commandArguments);
+        try {
+            final Process process;
+
+            process = new ProcessBuilder()
+                    .command(cmd)
+                    .start();
+
+            process.waitFor(10, TimeUnit.SECONDS);
```

Comment:
I think this should be removed here and leave to handle `Process` on calling method.